### PR TITLE
Various performance optimizations

### PR DESCRIPTION
**Pull request by smarter**

### CODE REVIEW

**Comment by smarter** [comment]
(I found this performance issue by using https://github.com/jvm-profiling-tools/async-profiler)

**Comment by smarter** [comment]
test performance please

**Comment by smarter** [comment]
I went ahead and did some more optimizations while I was at it :).

**Comment by smarter** [comment]
test performance please: 4bb0a1fe875ad64665ab0fd1a53104e52ebf56a7 5f0ed19d45818db33b31442bfc686b534dacb259 c47f0c244db1ed444adfd1f6d050b14a5645ccf9

File: compiler/src/dotty/tools/dotc/transform/TreeTransform.scala [hunk 1]
```diff
@@ -210,9 +210,16 @@ object TreeTransforms {
...
+    private def hasRedefinedMethod(cls: Class[_], name: String): Boolean = {
+      val clsMethods = cls.getDeclaredMethods
+      var i = clsMethods.length - 1
+      while (i >= 0) {
+        if (clsMethods(i).getName == name)
+          return cls != classOf[TreeTransform]
+        i -= 1
+      }
+      hasRedefinedMethod(cls.getSuperclass, name)
+    }
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/transform/TreeTransform.scala]: I wonder why this method shows up as hot. It should only be called during initialization when we compute the tree transformers. How often is the method called? Back of the envelope calculation: 100 phases * 50 methods, i.e. 5000? If its significantly more than that there's something we have overlooked.

**Review by odersky — APPROVED** [review]
It would be good to follow up on `hasRedefinedMethod`. Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/util/Stats.scala [hunk 2]
```diff
@@ -21,7 +21,7 @@ import collection.mutable
   }
 
   @inline
-  def record(fn: String, n: Int = 1) =
+  def record(fn: => String, n: => Int = 1) =
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/util/Stats.scala]: One can do even better by using two overloaded versions of `record` and no default argument.

Comment (smarter) [reply compiler/src/dotty/tools/dotc/util/Stats.scala]: I don't see how, could you give more details?

**Pull request merged**